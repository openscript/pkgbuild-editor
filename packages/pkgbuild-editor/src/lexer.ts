import { createToken, Lexer } from 'chevrotain';

export const Groups = {
  Formatting: createToken({ name: 'Formatting', pattern: Lexer.NA }),
};

export const Tokens = {
  Comment: createToken({ name: 'Comment', pattern: /#.*/ }),
  Newline: createToken({
    name: 'Newline',
    pattern: /\n/,
    line_breaks: true,
    categories: [Groups.Formatting],
  }),
  Whitespace: createToken({
    name: 'Whitespace',
    pattern: /\s+/,
    categories: [Groups.Formatting],
  }),
};

export const AllTokens = Object.values(Tokens);

export const PkgbuildLexer = new Lexer(AllTokens);
