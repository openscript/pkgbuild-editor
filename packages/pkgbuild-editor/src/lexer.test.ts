import { PkgbuildLexer } from './lexer';

describe('lexer.ts', () => {
  it('should tokenize a comment', () => {
    const lexingResult = PkgbuildLexer.tokenize(`# Test comment`);
    expect(lexingResult.tokens).toMatchSnapshot();
  });
  it('should tokenize multiple comments', () => {
    const lexingResult = PkgbuildLexer.tokenize(`# Test comment
# Another comment`);
    expect(lexingResult.tokens).toMatchSnapshot();
  });
});
