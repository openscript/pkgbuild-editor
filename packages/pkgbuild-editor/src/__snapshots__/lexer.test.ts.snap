// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lexer.ts > should tokenize a comment 1`] = `
[
  {
    "endColumn": 14,
    "endLine": 1,
    "endOffset": 13,
    "image": "# Test comment",
    "startColumn": 1,
    "startLine": 1,
    "startOffset": 0,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /#\\.\\*/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Comment",
      "tokenTypeIdx": 3,
    },
    "tokenTypeIdx": 3,
  },
]
`;

exports[`lexer.ts > should tokenize multiple comments 1`] = `
[
  {
    "endColumn": 14,
    "endLine": 1,
    "endOffset": 13,
    "image": "# Test comment",
    "startColumn": 1,
    "startLine": 1,
    "startOffset": 0,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /#\\.\\*/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Comment",
      "tokenTypeIdx": 4,
    },
    "tokenTypeIdx": 4,
  },
  {
    "endColumn": 15,
    "endLine": 1,
    "endOffset": 14,
    "image": "
",
    "startColumn": 15,
    "startLine": 1,
    "startOffset": 14,
    "tokenType": {
      "CATEGORIES": [
        {
          "CATEGORIES": [],
          "PATTERN": /NOT_APPLICABLE/,
          "categoryMatches": [
            5,
            6,
          ],
          "categoryMatchesMap": {
            "5": true,
            "6": true,
          },
          "isParent": true,
          "name": "Formatting",
          "tokenTypeIdx": 3,
        },
      ],
      "LINE_BREAKS": true,
      "PATTERN": /\\\\n/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Newline",
      "tokenTypeIdx": 5,
    },
    "tokenTypeIdx": 5,
  },
  {
    "endColumn": 17,
    "endLine": 2,
    "endOffset": 31,
    "image": "# Another comment",
    "startColumn": 1,
    "startLine": 2,
    "startOffset": 15,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /#\\.\\*/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Comment",
      "tokenTypeIdx": 4,
    },
    "tokenTypeIdx": 4,
  },
]
`;
